datafusion/core/src/physical_plan/sorts/sort.rs

```diff
@@ -124,6 +124,20 @@ impl ExternalSorter {
             // calls to `timer.done()` below.
             let _timer = tracking_metrics.elapsed_compute().timer();
             let partial = sort_batch(input, self.schema.clone(), &self.expr, self.fetch)?;
+            // The resulting batch might be smaller than the input batch if there
+            // is an propagated limit.
+            if self.fetch.is_some() {
+                let new_size = batch_byte_size(&partial.sorted_batch);
+                let size_delta = size.checked_sub(new_size).unwrap_or_else(|| {
+                    panic!(
```

Comment:
Would be better to return an internal error here instead of panicking.

Reply:
Makes sense, should be fixed now!